ethcontract/src/tokens.rs

```diff
@@ -0,0 +1,532 @@
+impl_single_tokenize_for_tuple!(12, A:0, B:1, C:2, D:3, E:4, F:5, G:6, H:7, I:8, J:9, K:10, L:11, );
+impl_single_tokenize_for_tuple!(13, A:0, B:1, C:2, D:3, E:4, F:5, G:6, H:7, I:8, J:9, K:10, L:11, M:12, );
+impl_single_tokenize_for_tuple!(14, A:0, B:1, C:2, D:3, E:4, F:5, G:6, H:7, I:8, J:9, K:10, L:11, M:12, N:13, );
+impl_single_tokenize_for_tuple!(15, A:0, B:1, C:2, D:3, E:4, F:5, G:6, H:7, I:8, J:9, K:10, L:11, M:12, N:13, O:14, );
+impl_single_tokenize_for_tuple!(16, A:0, B:1, C:2, D:3, E:4, F:5, G:6, H:7, I:8, J:9, K:10, L:11, M:12, N:13, O:14, P:15, );
+
+/// Rust type and multiple tokens conversion.
+pub trait MultiTokenize {
+    /// Convert self into tokens.
+    fn into_tokens(self) -> Vec<Token>;
+}
+
+/// Rust type and multiple token conversion.
+pub trait MultiDetokenize {
+    /// Convert tokens into Self.
+    fn from_tokens(tokens: Vec<Token>) -> Result<Self, Error>
+    where
+        Self: Sized;
+}
+
+macro_rules! impl_multi_for_tuple {
+  ($count: expr, $( $ty: ident : $no: tt, )*) => {
+    impl<$($ty, )*> MultiTokenize for ($($ty,)*) where
+      $(
+        $ty: SingleTokenize,
+      )*
+    {
+      fn into_tokens(self) -> Vec<Token> {
+        vec![
+          $( self.$no.into_token(), )*
+        ]
+      }
+    }
+
+    impl<$($ty, )*> MultiDetokenize for ($($ty,)*) where
+      $(
+        $ty: SingleTokenize,
+      )*
+    {
+      fn from_tokens(tokens: Vec<Token>) -> Result<Self, Error> {
+        if tokens.len() != $count {
+          return Err(Error::MultiDetokenizeLengthMismatch);
+        }
+        #[allow(unused_variables)]
+        #[allow(unused_mut)]
+        let mut drain = tokens.into_iter();
+        Ok(($(
+          $ty::from_token(drain.next().unwrap())?,
+        )*))
+      }
+    }
+  }
+}
+
+impl_multi_for_tuple!(0,);
+impl_multi_for_tuple!(1, A:0, );
+impl_multi_for_tuple!(2, A:0, B:1, );
+impl_multi_for_tuple!(3, A:0, B:1, C:2, );
+impl_multi_for_tuple!(4, A:0, B:1, C:2, D:3, );
+impl_multi_for_tuple!(5, A:0, B:1, C:2, D:3, E:4, );
+impl_multi_for_tuple!(6, A:0, B:1, C:2, D:3, E:4, F:5, );
+impl_multi_for_tuple!(7, A:0, B:1, C:2, D:3, E:4, F:5, G:6, );
+impl_multi_for_tuple!(8, A:0, B:1, C:2, D:3, E:4, F:5, G:6, H:7, );
+impl_multi_for_tuple!(9, A:0, B:1, C:2, D:3, E:4, F:5, G:6, H:7, I:8, );
+impl_multi_for_tuple!(10, A:0, B:1, C:2, D:3, E:4, F:5, G:6, H:7, I:8, J:9, );
+impl_multi_for_tuple!(11, A:0, B:1, C:2, D:3, E:4, F:5, G:6, H:7, I:8, J:9, K:10, );
+impl_multi_for_tuple!(12, A:0, B:1, C:2, D:3, E:4, F:5, G:6, H:7, I:8, J:9, K:10, L:11, );
+impl_multi_for_tuple!(13, A:0, B:1, C:2, D:3, E:4, F:5, G:6, H:7, I:8, J:9, K:10, L:11, M:12, );
+impl_multi_for_tuple!(14, A:0, B:1, C:2, D:3, E:4, F:5, G:6, H:7, I:8, J:9, K:10, L:11, M:12, N:13, );
+impl_multi_for_tuple!(15, A:0, B:1, C:2, D:3, E:4, F:5, G:6, H:7, I:8, J:9, K:10, L:11, M:12, N:13, O:14, );
+impl_multi_for_tuple!(16, A:0, B:1, C:2, D:3, E:4, F:5, G:6, H:7, I:8, J:9, K:10, L:11, M:12, N:13, O:14, P:15, );
```

Comment:
Do I understand correctly, that the problem with the Web3 Tokenize trade is this line:

> impl<T: Tokenizable> Tokenize for T

Since we don't implement this we don't get into the ambiguity that you mentioned? 

If that's the case maybe it would be better to try and leave the naming mostly similar and just remove this one implementation and add the `impl_single_tokenize_for_tuple` lines. This way the change is slightly easier to compare and might have a chance of being upstream compatible if we can get it merged upstream.

Reply:
Imo the web3 version and our version have different goals so I am unsure about whether we will go back to the web3 version. rust-web3 does not support a type safe api so it gives users more flexibility in the way they can provide rust types. We have full control over the rust interface so we can do some things differently and don't have to provide ergonomics.

Web3 *should* also add some kind of fix for tuples and I have made a PR for that but it is unclear whether that implementation would also work for ethcontract. That is why I treated this as more of  a from scratch kind of thing and renamed traits when I felt the old names were bad.

Reply:
I'd also be OK with moving `ethabi` to a `ethcontract-abi` and modifying it to better suit our needs. If I remember correctly, there are a few things that we end up having to work around because our needs don't fit very well with the library. Off the top of my head:
- Creating a lookup of [function name -> ABI fragment (this causes problems with overloaded function)](https://github.com/gnosis/ethcontract-rs/blob/8214590b791af0051c439918850247acdc098dd9/ethcontract/src/contract.rs#L318)
- for full v0.6 support we need to [pre-process the JSON](https://github.com/gnosis/ethcontract-rs/blob/8214590b791af0051c439918850247acdc098dd9/ethcontract-common/src/truffle.rs#L112)